### PR TITLE
Add debug adapter server

### PR DESCRIPTION
**Pull request by SpaceManiac**
Visual Studio Code extension: https://github.com/SpaceManiac/vscode-opendream
* Provides combined DM+C# debugging if your DM project and OpenDream checkout are both opened in a workspace together
* Otherwise allows you to configure a path to your OpenDream checkout and uses it in the background

Implemented:
* Basic setup (wixoa)
* Show `world.log` output in debug console (wixoa)
* Function breakpoints
* Break on exception
* Pause
* Continue
* Rudimentary threads, stack frames, scopes, and variables

Future work:
* Step in, step over, step out
* Line breakpoints

Architecture changes:
* DreamThread has a Name attribute
* Compiled `.json` includes source and line information for each proc's header
* ProcScheduler keeps track of sleeping procs so they can be discovered by the debugger

### CODE REVIEW

**Comment by ike709** [comment]
Holy crap

**Comment by Altoids1** [comment]
Wow!!

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Procs/DebugAdapter/DebugAdapterClient.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        _netStream = _client.GetStream();
+        _netReader = new StreamReader(_netStream, Encoding.ASCII);
+        _netWriter = new StreamWriter(_netStream, Encoding.ASCII);
+
+        _sawmill = Logger.GetSawmill("opendream.debug_adapter");
+        _sawmill.Info($"Accepted debug adapter client {_client.Client.RemoteEndPoint?.Serialize()}");
+    }
+
+    public void HandleMessages() {
+        while (_client.Connected && (_netStream.DataAvailable || _client.Available > 0 || _netReader.Peek() != -1)) {
```
[the remark this answers is not in the document]
Comment (SpaceManiac) [reply OpenDreamRuntime/Procs/DebugAdapter/DebugAdapterClient.cs:35]: `_netStream.DataAvailable` goes to false as soon as there is one Read call. `_client.Available` and `_netReader` each keep buffers and we have to loop until they are all drained.

File: OpenDreamRuntime/Procs/DebugAdapter/DebugAdapterClient.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        _client.Close();
+    }
+
+    public void SendMessage<T>(T message) where T : ProtocolMessage {
+        message.Seq = _seqCounter++;
+
+        string body = JsonSerializer.Serialize(message);
+        string header = $"Content-Length: {body.Length}\r\n\r\n";
+
+        _sawmill.Debug($"Sending {body.Length} {body}");
```
[the remark this answers is not in the document]
Comment (SpaceManiac) [reply OpenDreamRuntime/Procs/DebugAdapter/DebugAdapterClient.cs]: Switched to Verbose which is hidden by default, though I suppose they could also be totally removed

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**